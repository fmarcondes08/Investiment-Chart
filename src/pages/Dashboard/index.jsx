import React, { useState, useEffect, useCallback } from 'react';
import Chart from 'react-apexcharts';
import './styles.css';
import api from '../../services/api';
import { convertDate, getMonthFilter, getYearFilter } from '../../utils/utils';

function Dashboard() {
  const [investiment, setInvestiment] = useState([]);
  const [inputError, setInputError] = useState('');

  const options = {
    chart: {
      id: 'investimentChart',
      toolbar: {
        show: true,
      },
      zoom: {
        enabled: true,
      },
    },
    title: {
      text: 'Investments Chart',
      align: 'left',
    },
    subtitle: {
      text: 'Application Movement',
      align: 'left',
    },
    xaxis: {
      categories: investiment.map((x) => convertDate(x[0])),
    },
    yaxis: {
      opposite: true,
    },
  };

  const series = [
    {
      name: 'R$',
      data: investiment.map((x) => x[1]),
    },
  ];

  const fetchData = useCallback(
    async () => {
      try {
        const result = await api.get();
        setInvestiment(result.data);
        localStorage.setItem(
          '@InvestmentChart:investiments', JSON.stringify(result.data),
        );
      } catch (err) {
        // eslint-disable-next-line no-alert
        setInputError('Erro ao obter os dados, tente novamente.');
      }
    },
    [],
  );

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  function handleFilterInvestment(e) {
    const storageInvestiments = localStorage.getItem('@InvestmentChart:investiments')
      ? JSON.parse(localStorage.getItem('@InvestmentChart:investiments')) : '';

    if (!storageInvestiments) {
      setInputError('Erro ao obter os dados, tente novamente.');
      return '';
    }

    const lastMonth = storageInvestiments[storageInvestiments.length - 1];

    setInvestiment(() => {
      switch (e) {
        case '1': {
          const days = getMonthFilter(lastMonth[0], 1);
          const filter = storageInvestiments.filter((d) => d[0] >= days[0] && d[0] <= days[1]);
          return filter;
        }
        case '2': {
          const days = getMonthFilter(lastMonth[0], 3);
          const filter = storageInvestiments.filter((d) => d[0] >= days[0] && d[0] <= days[1]);
          return filter;
        }
        case '3': {
          const days = getYearFilter(lastMonth[0], 1);
          const filter = storageInvestiments.filter((d) => d[0] >= days[0] && d[0] <= days[1]);
          return filter;
        }
        case '4': {
          const days = getYearFilter(lastMonth[0], 2);
          const filter = storageInvestiments.filter((d) => d[0] >= days[0] && d[0] <= days[1]);
          return filter;
        }
        default: {
          return storageInvestiments;
        }
      }
    });

    return '';
  }

  return (
    <div className="container">
      {inputError && <span className="error">{inputError}</span>}
      <div className="filter">
        <select
          id="filterInvestiment"
          className="filterInvestiment"
          autoFocus
          onChange={(e) => handleFilterInvestment(e.target.value)}
        >
          <option value="0">From the beginning</option>
          <option value="1">Last Month</option>
          <option value="2">3 Months</option>
          <option value="3">1 Year</option>
          <option value="4">2 Years</option>
        </select>
      </div>
      <div id="chart" className="chart">
        <Chart options={options} series={series} type="area" width="1000" height="500" />
      </div>
    </div>
  );
}

export default Dashboard;
