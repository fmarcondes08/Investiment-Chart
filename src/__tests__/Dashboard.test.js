import React from 'react';
import {
  render, fireEvent, cleanup,
} from '@testing-library/react';
import Dashboard from '../pages/Dashboard/index';


afterEach(cleanup);

jest.mock('axios');

export default {
  get: jest.fn(() => Promise.resolve({ data: {} })),
};

jest.mock('react-apexcharts', () => function DummyChart(series, options) {
  return (
    <div data-testid="chart" series={series} options={options} />
  );
});

describe('Dashboard component', () => {
  const options = {
    chart: {
      id: 'investimentChart',
      toolbar: {
        show: true,
      },
      zoom: {
        enabled: true,
      },
    },
    title: {
      text: 'Investments Chart',
      align: 'left',
    },
    subtitle: {
      text: 'Application Movement',
      align: 'left',
    },
    xaxis: {
      categories: [],
    },
    yaxis: {
      opposite: true,
    },
  };

  const series = [
    {
      name: 'R$',
      data: [],
    },
  ];

  it('should take a snapshot', () => {
    const { asFragment } = render(<Dashboard series={series} options={options} />);

    expect(asFragment(<Dashboard />)).toMatchSnapshot();
  });

  it('Should change select', () => {
    const { getByDisplayValue } = render(<Dashboard series={series} options={options} />);

    const select = getByDisplayValue(/From the beginning/i);
    fireEvent.change(select, { target: { value: '1' } });

    expect(getByDisplayValue(/Last Month/i)).toBeInTheDocument();
  });
});
