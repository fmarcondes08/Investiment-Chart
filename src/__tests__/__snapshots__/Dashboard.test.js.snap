// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <span
      class="error"
    >
      Erro ao obter os dados, tente novamente.
    </span>
    <div
      class="filter"
    >
      <select
        class="filterInvestiment"
        id="filterInvestiment"
      >
        <option
          value="0"
        >
          From the beginning
        </option>
        <option
          value="1"
        >
          Last Month
        </option>
        <option
          value="2"
        >
          3 Months
        </option>
        <option
          value="3"
        >
          1 Year
        </option>
        <option
          value="4"
        >
          2 Years
        </option>
      </select>
    </div>
    <div
      class="chart"
      id="chart"
    >
      <div
        data-testid="chart"
        options="[object Object]"
        series="[object Object]"
      />
    </div>
  </div>
</DocumentFragment>
`;
